src/doppler-types.ts

```diff
@@ -64,39 +64,210 @@ export interface PrepaidPack {
+    | MonthlyRenewalDeliveriesPlan
+    | SubscribersLimitedPlan
+  )[];
+
+const getPlusPlans = (
+  planList: Plan[],
+): (MonthlyRenewalDeliveriesPlan | SubscribersLimitedPlan)[] =>
+  planList.filter((x) => x.featureSet == 'plus') as (
+    | MonthlyRenewalDeliveriesPlan
+    | SubscribersLimitedPlan
+  )[];
+
+const getFreePathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [FreePath] => {
+  var plans = getFreePlans(planList);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'free',
+      actual: userPlan.featureSet == 'free',
+      deadEnd: plans.length == 1 && userPlan == cheapestPlan,
+    },
+  ];
+};
+
+const getAgencyPathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [AgenciesPath] => {
+  var plans = getAgencyPlans(planList);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'agencies',
+      actual: userPlan.featureSet == 'agency',
+      deadEnd: plans.length == 1 && userPlan == cheapestPlan,
+    },
+  ];
+};
+
+const getStandardPathOrEmpty = (userPlan: Plan, planList: Plan[]): [] | [StandardPath] => {
+  var plans = getStandardPlans(planList).sort((x) => x.fee);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'standard',
+      actual: userPlan.featureSet == 'standard',
+      minimumFee: cheapestPlan.fee,
+      deadEnd: plans.length == 1 && cheapestPlan == userPlan,
+    },
+  ];
+};
+
+const getPlusPathOrEmpty = (userPlan: Plan, planList: Plan[]): PlusPath[] => {
+  var plans = getPlusPlans(planList).sort((x) => x.fee);
+
+  if (plans.length == 0) {
+    return [];
+  }
+
+  var cheapestPlan = plans[0];
+
+  return [
+    {
+      type: 'plus',
+      actual: userPlan.featureSet == 'plus',
+      minimumFee: cheapestPlan.fee,
+      deadEnd: plans.length == 1 && cheapestPlan == userPlan,
+    },
+  ];
+};
+
+const _agencyPlan: AgencyPlan = {
+  type: 'agency',
+  featureSet: 'agency',
+};
+
+export const getPaths = (userPlan: Plan, planList: Plan[]): Path[] => {
+  // Assuming that planList does not contains 'free' and neither 'agency'
+  const potentialUpgradePlans =
+    userPlan.type == 'free'
+      ? planList
+      : userPlan.type == 'prepaid'
+      ? [
+          ...getPrepaidPacks(planList),
+          ...getUpgradeMonthlyPlans(planList),
+          ...getUpgradeSubscribersPlans(planList),
+        ]
+      : userPlan.type == 'monthly-deliveries'
+      ? getUpgradeMonthlyPlans(planList, {
+          minFee: userPlan.fee,
+          minEmailsByMonth: userPlan.emailsByMonth,
+        })
+      : userPlan.type == 'subscribers'
+      ? getUpgradeSubscribersPlans(planList, {
+          minFee: userPlan.fee,
+          minSubscriberLimit: userPlan.subscriberLimit,
+        })
+      : [];
+
+  const potentialAndCurrentPlans = [userPlan, ...potentialUpgradePlans, _agencyPlan];
```

Comment:
And here I should add free plan and agency, because they don´t exist in BE.

Reply:
Well, in fact, doing that is not necessary. 

`potentialUpgradePlans` will never include the `free plan`, so `potentialAndCurrentPlans` will only include the `free plan` when the `userPlan` is the `userPlan`.

On the other hand, `_agencyPlan` is always a `potentialUpgradePlans` and it is not included in the `potentialUpgradePlans`(because they don't exist in BE), so adding it here is the right thing.

---

Well, in fact, there is a broken edge case:

When the `userPlan` is the `_agencyPlan`, `potentialAndCurrentPlans` will include the `_agencyPlan` twice.